src/front/glsl/builtins.rs

```diff
@@ -0,0 +1,1650 @@
+                        _ => Some(VectorSize::Quad),
+                    };
+
+                    let ty = || match size {
+                        Some(size) => TypeInner::Vector {
+                            size,
+                            kind: Sk::Float,
+                            width,
+                        },
+                        None => TypeInner::Scalar {
+                            kind: Sk::Float,
+                            width,
+                        },
+                    };
+                    let args = vec![
+                        ty(),
+                        ty(),
+                        TypeInner::Scalar {
+                            kind: Sk::Float,
+                            width: 4,
+                        },
+                    ];
+                    declarations.push(add_builtin(
+                        &mut self.module,
+                        args,
+                        Builtins::MathFunction(MathFunction::Refract),
+                    ))
+                }
+            }
+            "smoothstep" => {
+                // bits layout
+                // bit 0 - float/double
+                // bit 1 - splatted
+                // bit 2 trough 3 - dims
+                for bits in 0..(0b10000) {
+                    let width = (bits & 0b1) * 4 + 4;
+                    let splatted = bits & 0b10 == 0b10;
+                    let size = match bits >> 2 {
+                        0b00 => None,
+                        0b01 => Some(VectorSize::Bi),
+                        0b10 => Some(VectorSize::Tri),
+                        _ => Some(VectorSize::Quad),
+                    };
+
+                    if splatted && size.is_none() {
+                        continue;
+                    }
+
+                    let base_ty = || match size {
+                        Some(size) => TypeInner::Vector {
+                            size,
+                            kind: Sk::Float,
+                            width,
+                        },
+                        None => TypeInner::Scalar {
+                            kind: Sk::Float,
+                            width,
+                        },
+                    };
+                    let ty = || match splatted {
+                        true => TypeInner::Scalar {
+                            kind: Sk::Float,
+                            width,
+                        },
+                        false => base_ty(),
+                    };
+                    declarations.push(add_builtin(
+                        &mut self.module,
+                        vec![ty(), ty(), base_ty()],
+                        Builtins::MathFunction(MathFunction::SmoothStep),
+                    ))
+                }
+            }
+            // Not a builtin (or at lest not yet handled)
+            _ => {}
+        }
+    }
+}
+
+/// A compiler defined builtin function
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub enum Builtins {
```

Comment:
the name should be singular (`Builtin`) or better something that doesn't conflict with IR names